### PR TITLE
More Coder doc

### PR DESCRIPTION
**Pull request by jto**

### CODE REVIEW

File: site/src/paradox/internals/Coders.md [hunk 1]
```diff
@@ -8,9 +8,10 @@ As per [Beam's documentation](https://beam.apache.org/documentation/programming-
 
 > When Beam runners execute your pipeline, they often need to materialize the intermediate data in your PCollections, which requires converting elements to and from byte strings. The Beam SDKs use objects called Coders to describe how the elements of a given PCollection may be encoded and decoded.
 
-So anytime you create a `SCollection[T]`, Beam needs to know how to go from an instance of `T` to an array of bytes, and from that array of bytes to an instance of `T`.
+For the most part, coders are used when Beam transfer intermediate data between workers over the network. They may also be used by beam to test instances for equality.
+Anytime you create a `SCollection[T]`, Beam needs to know how to go from an instance of `T` to an array of bytes, and from that array of bytes to an instance of `T`.
 
-The Beam SDK has a class called `Coder` that roughly looks like this:
+The Beam SDK defines class called `Coder` that roughly looks like this:
```
Comment (regadas) [inline site/src/paradox/internals/Coders.md:14]:
```suggestion
The Beam SDK defines a class called `Coder` that roughly looks like this:
```

File: site/src/paradox/internals/Coders.md [hunk 2]
```diff
@@ -25,7 +26,56 @@ Beam provides built-in Coders for various basic Java types (`Integer`, `Long`, `
 import com.spotify.scio.values.SCollection
 
 case class Foo(x: Int, s: String)
-def sc: SCollection[Foo] = ??? // Beam will need an org.apache.beam.sdk.coders.Coder[Foo]
+def coll: SCollection[Foo] = ??? // Beam will need an org.apache.beam.sdk.coders.Coder[Foo]
+```
+
+### When are `Coder` used ?
```
Comment (regadas) [inline site/src/paradox/internals/Coders.md:32]: perhaps?

```suggestion
### When are `Coder`'s used ?
```

File: site/src/paradox/internals/Coders.md [hunk 3]
```diff
@@ -25,7 +26,56 @@ Beam provides built-in Coders for various basic Java types (`Integer`, `Long`, `
...
+coder.encode(id2) // 00010010
+// -> i1 and i2 are NOT equal
+```
+
+**When they are used to test equality, coders are required to be deterministic.**
+If a non-deterministic `Coder` is used to test equality, an exception is thrown:
+
+```scala mdoc:reset:invisible
+import com.spotify.scio.ScioContext
+val sc = ScioContext.forTest()
```
Comment (regadas) [inline site/src/paradox/internals/Coders.md:71]: should we name this more explicit? perhaps `context` or `scioContext`?

File: site/src/paradox/internals/Coders.md [hunk 4]
```diff
@@ -41,31 +91,69 @@ While it saves a lot of work, it also has a few drawbacks:
...
+Scio `Coder` and its implementations simply form an [ADT](https://en.wikipedia.org/wiki/Algebraic_data_type) where each implementation is a building block that covers one of the possible cases:
+
+- `Beam`: a simple wrapper around a Beam Coder
+- `Disjunction`: Represent a Coder that makes a choice between different possible implementations. It is for example used to serialize ADTs and `Either`
+- `Record`: A Coder for record-like structures like case classes and tuples.
+- `Transform`: A Coder implemented by "transforming" another Coder.
+- `Fallback`: A default `Coder`. Used when there is no better option.
 
-Please note that Scio wraps Beam coders in its own `Coder` definition: `com.spotify.scio.coders.Coder`
+There is also a "special" coder called `KVCoder`. It is a specific coder for Key-Value pairs. Internally Beam treats [KV](https://beam.apache.org/releases/javadoc/2.20.0/org/apache/beam/sdk/values/KV.html) differently from other types so Scio needs to do the same.
```
Comment (regadas) [inline site/src/paradox/internals/Coders.md:108]: I think we could use the `@javadoc[KV](org.apache.beam.sdk.values.KV)` or did it not work in this case? advantage is that points to whatever beam version we are using

File: site/src/paradox/internals/Coders.md [hunk 5]
```diff
@@ -138,7 +227,93 @@ case class Demo2(i: Int, s: String, xs: List[java.util.Locale])
...
+```
+- **`Coder.transform`**: Create a Coder for a type `B` by transforming the Beam implementation for a type `A`. Usually useful for `Coder` that depend on another `Coder`:
+```scala mdoc
+import java.io.{InputStream, OutputStream}
+import org.apache.beam.sdk.coders.AtomicCoder
+class ListCoder[T](bc: org.apache.beam.sdk.coders.Coder[T]) extends AtomicCoder[List[T]] {
+  override def encode(value: List[T], outStream: OutputStream): Unit = ???
+  override def decode(inStream: InputStream): List[T] = ???
+}
+implicit def listCoder[T: Coder]: Coder[List[T]] = Coder.transform(Coder[T])(bc => Coder.beam(new ListCoder[T](bc)))
```
Comment (regadas) [inline site/src/paradox/internals/Coders.md:278]: should we rename `bc` for clarity?

File: site/src/paradox/internals/Coders.md (hunk 2 above)
Comment (jto) [reply site/src/paradox/internals/Coders.md:32]: Uh ? No ?

File: site/src/paradox/internals/Coders.md (hunk 3 above)
Comment (jto) [reply site/src/paradox/internals/Coders.md:71]: We call it `sc` pretty much everywhere in docs and example. I agree it's no ideal but consistency is nice ^^

File: site/src/paradox/internals/Coders.md (hunk 4 above)
Comment (jto) [reply site/src/paradox/internals/Coders.md:108]: I did not know `@javadoc` existed :/

**Pull request merged**